pkg/services/dashboards/service/zanzana.go

```diff
@@ -0,0 +1,325 @@
+	}
+	return dr.findDashboardsZanzanaCheck
+}
+
+// findDashboardsZanzanaCheck implements "Search, then check" strategy. It first performs search query, then filters out results
+// by checking access to each item.
+func (dr *DashboardServiceImpl) findDashboardsZanzanaCheck(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
+	ctx, span := tracer.Start(ctx, "dashboards.service.findDashboardsZanzanaCheck")
+	defer span.End()
+
+	result := make([]dashboards.DashboardSearchProjection, 0)
+	var page int64 = 1
+	query.SkipAccessControlFilter = true
+	// Set limit to default to prevent pagination issues
+	limit := query.Limit
+
+	for len(result) < int(limit) {
+		query.Page = page
+		query.Limit = defaultQueryLimit
+		findRes, err := dr.dashboardStore.FindDashboards(ctx, query)
+		if err != nil {
+			return nil, err
+		}
+
+		query.Limit = limit
+		res, err := dr.checkDashboards(ctx, query, findRes)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, res...)
+		page++
+
+		// Stop when last page reached
+		if len(findRes) < defaultQueryLimit {
+			break
+		}
+	}
+
+	if len(result) > int(limit) {
+		result = result[:limit]
+	}
+
+	return result, nil
+}
+
+func (dr *DashboardServiceImpl) checkDashboards(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery, searchRes []dashboards.DashboardSearchProjection) ([]dashboards.DashboardSearchProjection, error) {
+	ctx, span := tracer.Start(ctx, "dashboards.service.checkDashboards")
+	defer span.End()
+
+	if len(searchRes) == 0 {
+		return nil, nil
+	}
+
+	orgId := query.OrgId
+	if orgId == 0 && query.SignedInUser.GetOrgID() != 0 {
+		orgId = query.SignedInUser.GetOrgID()
+	}
+
+	concurrentRequests := dr.cfg.Zanzana.ConcurrentChecks
+	var wg sync.WaitGroup
+	res := make([]dashboards.DashboardSearchProjection, 0)
+	resToCheck := make(chan dashboards.DashboardSearchProjection, concurrentRequests)
+	allowedResults := make(chan dashboards.DashboardSearchProjection, len(searchRes))
+
+	for i := 0; i < int(concurrentRequests); i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for d := range resToCheck {
+				if len(allowedResults) >= int(query.Limit) {
+					return
+				}
+
+				objectType := zanzana.TypeDashboard
+				if d.IsFolder {
+					objectType = zanzana.TypeFolder
+				}
+
+				req := accesscontrol.CheckRequest{
+					User:     query.SignedInUser.GetUID(),
+					Relation: "read",
+					Object:   zanzana.NewScopedTupleEntry(objectType, d.UID, "", strconv.FormatInt(orgId, 10)),
+				}
+
+				allowed, err := dr.ac.Check(ctx, req)
+				if err != nil {
+					dr.log.Error("error checking access", "error", err)
+				} else if allowed {
+					allowedResults <- d
+				}
+			}
+		}()
+	}
+
+	for _, r := range searchRes {
+		resToCheck <- r
+	}
+	close(resToCheck)
+
+	wg.Wait()
+	close(allowedResults)
+
+	for r := range allowedResults {
+		if len(res) >= int(query.Limit) {
+			break
+		}
```

Comment:
Good catch, I'd have forgotten this one.
```suggestion
		if int64(len(res)) >= remains {
			break
		}
```